File: packages/shared-ts/src/session-secret-store.ts
export interface SessionSecretRecord {
  name: string;
  value: string;
  domains: string[];
}

export class SessionSecretStore {
  private readonly entries = new Map<string, SessionSecretRecord>();

  set(name: string, value: string, domains: string[] = []): void {
    this.entries.set(name, { name, value, domains: [...domains] });
  }

  get(name: string): string | undefined {
    return this.entries.get(name)?.value;
  }

  getRecord(name: string): SessionSecretRecord | undefined {
    const r = this.entries.get(name);
    return r ? { name: r.name, value: r.value, domains: [...r.domains] } : undefined;
  }

  has(name: string): boolean {
    return this.entries.has(name);
  }

  setDomains(name: string, domains: string[]): boolean {
    const r = this.entries.get(name);
    if (!r) return false;
    r.domains = [...domains];
    return true;
  }

  delete(name: string): boolean {
    return this.entries.delete(name);
  }

  listAll(): SessionSecretRecord[] {
    return Array.from(this.entries.values()).map((r) => ({
      name: r.name,
      value: r.value,
      domains: [...r.domains],
    }));
  }

  list(): Array<{ name: string; domains: string[] }> {
    return Array.from(this.entries.values()).map((r) => ({
      name: r.name,
      domains: [...r.domains],
    }));
  }

  size(): number {
    return this.entries.size;
  }
}

export function previewSecret(value: string, edge = 4): string {
  const len = value.length;
  if (len === 0) return '';
  if (len <= 2) return '…';
  const e = Math.min(Math.max(1, edge), Math.floor((len - 1) / 2));
  return `${value.slice(0, e)}…${value.slice(len - e)}`;
}

File: packages/swift-server/Sources/Keychain/SessionSecretStore.swift
import Foundation

/// Process-memory-only session secrets. Values are never persisted to Keychain
/// or disk and disappear when the server process exits.
actor SessionSecretStore {
    struct Record: Sendable, Equatable {
        let name: String
        let value: String
        let domains: [String]
    }

    private var entries: [String: Record] = [:]

    func set(name: String, value: String, domains: [String] = []) {
        entries[name] = Record(name: name, value: value, domains: domains)
    }

    func getRecord(name: String) -> Record? {
        entries[name]
    }

    func setDomains(name: String, domains: [String]) -> Bool {
        guard let existing = entries[name] else { return false }
        entries[name] = Record(name: name, value: existing.value, domains: domains)
        return true
    }

    func delete(name: String) -> Bool {
        entries.removeValue(forKey: name) != nil
    }

    func listAll() -> [Record] {
        Array(entries.values)
    }

    func list() -> [SecretEntry] {
        entries.values.map { SecretEntry(name: $0.name, domains: $0.domains) }
    }
}

/// Return a partial preview while always eliding at least one character.
func previewSecret(_ value: String, edge: Int = 4) -> String {
    let characters = Array(value)
    guard !characters.isEmpty else { return "" }
    guard characters.count > 2 else { return "…" }
    let visibleEdge = min(max(1, edge), (characters.count - 1) / 2)
    return String(characters.prefix(visibleEdge)) + "…" + String(characters.suffix(visibleEdge))
}
